Step1, Step2: 지하철 노선도

ProtoSeo opened the pull request:
안녕하세요. 우아한테크캠프 6기 서승훈입니다.
세 번째 미션 지하철 노선의 1단계 미션, 2단계 미션에 대한 Pull request입니다.
다음은 페어와 함께 이번 미션을 진행하면서 겪었던 문제와 고민입니다.

## 미션을 진행하며
미션을 시작하면서 설계를 할 때 데이터베이스의 id로 값을 가져오는 것이 좋다고 판단하였습니다. 그래서 데이터베이스의 관계로 도메인의 관계를 정의하였습니다. 뒤늦게 강의를 들으면서 이것이 간접 참조 방식임을 알게 되었습니다.

하지만 이렇게 개발하면서 이전 미션에서 배웠던 객체 지향 프로그래밍을 적용하기 어렵다고 느꼈습니다. 레이어 간에 서로 침범한다는 생각도 들었고, 도메인 객체가 데이터베이스에 의존적이라는 생각도 들었습니다.

그래서 2단계 기능 구현을 마치고 되돌아 봤을 때, 도메인과 엔티티가 분리되지 않아서 이런 문제가 발생한 것 같다고 생각했습니다. 도메인과 엔티티의 개념이 혼재돼어 DAO가 너무 많은 일을 하는 것이 아닌가 의문이 들었습니다. 마치 DAO가 도메인을 생성하는것처럼 보이게 됩니다. 이를 위해 Repository 레이어를 추가했으면 어땠을까 하는 생각이 있습니다.

## 궁금한 점
* 도메인과 엔티티에 관해 다음과 같이 생각했습니다. 저희가 생각하는 것이 맞는 지 궁금합니다.
    * 도메인: 비즈니스 로직을 처리하는 단위라고 생각했습니다.
    * 엔티티: DB 쿼리 결과 컬럼의 집합을 표현하는 단위라고 생각했습니다.
* 2단계 기능 구현을 마치고 되돌아 봤을 때 일련의 고민들이 도메인과 엔티티가 분리되지 않아서 문제가 발생한 것 같습니다. 제가 생각한 것이 맞을까요?

* DAO는 DB를 직접적으로 불러오는 낮은 단계의 추상화 레이어이고, Repository는 반대로 높은 단계의 추상화 레벨이라고 생각합니다. 만약 도메인과 엔티티를 분리한다면 엔티티는 DAO로, 도메인은 Repository로 관리하는 것이 좋을까요? 아니면 다른 방법이 있을까요?

* TDD를 한다고 열심히 하긴 했는데 TDD로 통합 테스트를 진행하다보니 너무 큰 단위를 테스트한다는 느낌이 들었습니다. 지금처럼 계층이 분리된 미션을 할 때 어떻게 TDD를 하는 것이 좋을까요? 작은 단위에서 쌓아서 올라가는 것이 좋을까요? 아니면 그 반대가 좋을까요? TDD를 진행하면서 많은 혼란을 겪었습니다.

whatasame committed:
docs: 페어 프로그래밍 규칙 작성

Co-authored-by: protoseo <proto_seo@naver.com>

whatasame committed:
docs: 1단계 요구 사항 작성

Co-authored-by: protoseo <proto_seo@naver.com>

whatasame committed:
feat: 구간 도메인 정의

Co-authored-by: protoseo <proto_seo@naver.com>

whatasame committed:
docs: 구간 테이블 스키마 정의

Co-authored-by: protoseo <proto_seo@naver.com>

whatasame committed:
fix: 직렬화를 위한 기본 생성자 추가

Co-authored-by: protoseo <proto_seo@naver.com>

whatasame committed:
docs: 누락한 distance 컬럼 추가

Co-authored-by: protoseo <proto_seo@naver.com>

whatasame committed:
feat: SectionDao 추가

Co-authored-by: protoseo <proto_seo@naver.com>

whatasame committed:
test: JDBC 동작 확인 테스트

Co-authored-by: protoseo <proto_seo@naver.com>

whatasame committed:
docs: 기본 DML 파일명으로 변경

Co-authored-by: protoseo <proto_seo@naver.com>

whatasame committed:
fix: 디버그용 toString() 제거

Co-authored-by: protoseo <proto_seo@naver.com>

whatasame committed:
docs: 구간 테이블의 외래키 추가

Co-authored-by: protoseo <proto_seo@naver.com>

whatasame committed:
test: 테스트 케이스 수정 및 추가

Co-authored-by: protoseo <proto_seo@naver.com>

whatasame committed:
feat: CORS 처리 추가

Co-authored-by: protoseo <proto_seo@naver.com>

whatasame committed:
docs: 불필요한 요구 사항 삭제

Co-authored-by: protoseo <proto_seo@naver.com>

whatasame committed:
refactor: DB 컬럼명 수정

Co-authored-by: protoseo <proto_seo@naver.com>

whatasame committed:
feat: 구간 등록 기능 구현

Co-authored-by: protoseo <proto_seo@naver.com>

whatasame committed:
refactor: 코드 컨벤션 수정

Co-authored-by: protoseo <proto_seo@naver.com>

whatasame committed:
feat: 새로운 구간의 하행역이 해당 노선에 등록되어 있는 경우 400 Bad Request로 응답

Co-authored-by: protoseo <proto_seo@naver.com>

whatasame committed:
docs: 테스트 데이터 추가

Co-authored-by: protoseo <proto_seo@naver.com>

whatasame committed:
test: 새로운 구간의 상행역이 해당 노선에 등록되어 있는 하행 종점역이 아닌 경우 400 Bad Request로 응답

Co-authored-by: protoseo <proto_seo@naver.com>

whatasame committed:
feat: Section 관련 기능 추가

Co-authored-by: protoseo <proto_seo@naver.com>

whatasame committed:
feat: Section들을 관리하는 일급 컬렉션 정의

Co-authored-by: protoseo <proto_seo@naver.com>

whatasame committed:
feat: 새로운 구간의 상행역이 해당 노선에 등록되어 있지 않은 경우 400 Bad Request로 응답

Co-authored-by: protoseo <proto_seo@naver.com>

whatasame committed:
docs: 테스트 데이터 수정

Co-authored-by: protoseo <proto_seo@naver.com>

whatasame committed:
feat: 구간 제거 기능 구현

Co-authored-by: protoseo <proto_seo@naver.com>

whatasame committed:
refactor: 누락된 final 추가

Co-authored-by: protoseo <proto_seo@naver.com>

whatasame committed:
refactor: 코드 리팩토링

- 누락된 final 추가
- updatePrevSectionId 테스트 추가
- lineId query param 추출
- 테스트에서만 사용하는 findById 메서드 삭제

Co-authored-by: protoseo <proto_seo@naver.com>

whatasame committed:
fix: NULL 조회는 `is NULL`로 조회

Co-authored-by: protoseo <proto_seo@naver.com>

whatasame committed:
feat: 지하철 노선에 등록된 하행 종점역이 아닌 경우 삭제 시 400 Bad Request로 응답

Co-authored-by: protoseo <proto_seo@naver.com>

whatasame committed:
feat: 지하철 노선에 구간이 하나인 경우 삭제 시 400 Bad Request로 응답

Co-authored-by: protoseo <proto_seo@naver.com>

whatasame committed:
docs: 1단계 미션 완료

Co-authored-by: protoseo <proto_seo@naver.com>

whatasame committed:
fix: 잘못된 메서드명 수정

Co-authored-by: protoseo <proto_seo@naver.com>

whatasame committed:
refactor: 구간 스키마 수정

Co-authored-by: protoseo <proto_seo@naver.com>

whatasame committed:
docs: 2단계 요구 사항 작성

Co-authored-by: protoseo <proto_seo@naver.com>

whatasame committed:
feat: 상행 역과 하행 역이 이미 노선에 모두 등록되어 있는 경우 400 Bad Request로 응답

Co-authored-by: protoseo <proto_seo@naver.com>

whatasame committed:
feat: 상행 역과 하행 역이 모두 노선에 없는 경우 400 Bad Request로 응답

Co-authored-by: protoseo <proto_seo@naver.com>

whatasame committed:
feat: 새로운 역을 상행 종점역이나 하행 종점역으로 등록

Co-authored-by: protoseo <proto_seo@naver.com>

whatasame committed:
refactor: 누락된 final 추가

Co-authored-by: protoseo <proto_seo@naver.com>

whatasame committed:
feat: 구간 추가 기능 심화 구현

Co-authored-by: protoseo <proto_seo@naver.com>

whatasame committed:
feat: 리액트 대응

Co-authored-by: protoseo <proto_seo@naver.com>

sihyung92 commented on README.md:
```diff
@@ -1 +1,63 @@
+
+## 페어 프로그래밍
+
+### 목표
+
+* 서승훈: 즐기면서 하자
+* 최정규: 모든걸 의심하자
```
ㅋㅋ 재밌네요 안 싸우셨나요?

ProtoSeo replied:
ㅎㅎ 서로 재미있게 잘 진행했습니다!

sihyung92 commented on README.md:
```diff
@@ -1 +1,63 @@
+* 잡담을 많이 하자
+* 시간적 압박 등의 부담 없이 하자
+* 너무 완벽하려고 하지 말자
+* 고민이 길어지면 가위바위보로 정하자
```
사다리 타기도 추천드립니다. 우주최고존엄 의사결정기 입니다. 
https://www.youtube.com/watch?v=BRaNXWYKDuo
은근히 나에게 유리한 선택지를 유도할 수 있다는 것도 장점입니다.

ProtoSeo replied:
오 사다리타기는 생각하지 못했었는데, 좋은 방법인 것 같아요! 다음에 꼭 사용해보도록 하겠습니다~

sihyung92 commented on src/test/java/subway/dao/SectionDaoTest.java:
```diff
@@ -0,0 +1,44 @@
+        assertThat(insert).isEqualTo(section);
+    }
+
+    @Test
+    @DisplayName("구간을 하나 삭제한다.")
+    void delete() {
+        /* given */
+        final Long targetSectionId = 3L;
+
+        /* when */
+        final int result = sectionDao.delete(targetSectionId);
+
+        /* then */
+        assertThat(result).isEqualTo(1);
+    }
```
테스트 대상에  대한 세팅이 운영 프로덕션 세팅 데이터인 data.sql에 의존해서 이루어지고 있습니다. 이 경우 data.sql의 변경사항에 대해 테스트가 취약해지므로, 테스트용 데이터 생성과 유지보수 가능한 테스트에 대해 고민해보셔요.

sihyung92 replied:
그리고 테스트 상 매직넘버가 많아 테스트를 파악하기 어렵습니다. SectionTest의 경우 상수를 생성해 가독성이 훌륭한 테스트를 작성해주셨는데, 이쪽 코드도 마찬가지로 리팩토링 해주세요. 유지보수가 어려운 테스트를 작성한 경우 휴가를 가기 어렵습니다 (테스트 깨질 경우 전화오거나 테스트 삭제당함)

ProtoSeo replied:
휴가를 가지 못 할수 있다니,, 테스트코드도 잘 작성하는 것이 중요하겠네요! 

먼저 `data.sql`은 원래는 테스트를 위한 데이터 더미였는데 잠시 `react-subway`를 위해 옮겼던 것을 다시 `test`로 이동시켰습니다. 
그리고 `SectionTest`, `SectionDaoTest`의 가독성을 매직넘버를 제거하고, 공통적으로 사용하는 부분은 매개변수로 사용해서 최대한 가독성을 좋게 만들려고 코드를 수정해봤습니다!

ProtoSeo replied:
```java
final static Sections SECTIONS = new Sections(List.of(
            new Section(1L, LINE_ID, 1L, 2L, 1L),
            new Section(2L, LINE_ID, 2L, 3L, 2L),
            new Section(3L, LINE_ID, 3L, 4L, 3L),
            new Section(4L, LINE_ID, 4L, 5L, 4L)
    ));
```

`SectionsTest`에서 `SECTIONS`를 생성할 때 사용하는 `upStationId`, `downStationId` 같은 ID값도 매직넘버 제거할 때 같이 제거하고 싶은데, 각각을 상수로 빼버리면 너무 많은 상수가 생성될 것 같아서 그냥 내비 두었는데요 이런 부분은 어떻게 처리하는 것이 좋을까요?

- 그냥 아이디 값(`1L`, `2L`, ...)과 같이 둬도 가독성에 문제가 없다.
- 아이디 값들도 싱수로 뺴서 일단 명확하게 표현을 해주는 것이 좋다.

ProtoSeo replied:
또 질문이 있는데요, 따로 이런 테스트 객체만 생성하는 클래스를 만들어서 (예를 들어 `SectionFixture`, `StationFixture`와 같이?) 관리하는 방법도 생각해보긴 했는데요, 테스트에서 사용하는 객체들이 그렇게 많지는 않은 것 같아서, 생성해서 관리하지는 않았는데, 차라리 Fixture로 관리하는 것이 나았을까요?

sihyung92 replied:
상수값은 제삼자의 눈으로 봤을 때 가장 읽기 좋은 방식으로 해주세요. 
모든 원시값을 상수화해도 읽기 좋지 않고, 지금처럼 원시값밖에 좋지하지 않아도 읽기 좋지 않습니다. 저는 보통 테스트의 비교군이 되는 값들만 상수화 해주어요.

예컨데 다음 문장은 눈에 잘 들어오지 않습니다.


```java
        assertThat(SECTIONS.containsBoth(4L, 5L)).isTrue();
```

4L, 5L의 Section을 미리 상수화 해두면 다음처럼 볼 수 있습니다.

```java
        assertThat(
         SECTIONS.containsBoth(
		산본역_금정역_구간.getUpStationId(),
		산본역_금정역_구간.getDownStationId()
         )
      ).isTrue();
```

sihyung92 replied:
Fixture도 관리하기 쉬운 형태로 하시면 될거같아요. 컨벤션을 정했다면 컨벤션대로, 그렇지 않다면 저같은 경우엔 간단하고 해당 테스트 파일에서만 쓰일거 같은건 inner class나 static method를 통해 표현도 자주 합니다. 다른 곳에서 쓰일거 같다면 따로 클스로 빼두면 나중에 다른 테스트에서도 용이하게 쓰이기도 하고요.

sihyung92 commented on src/main/java/subway/dto/StationResponse.java:
```diff
@@ -6,6 +6,10 @@ public class StationResponse {
     private Long id;
     private String name;
 
+    private StationResponse() {
+        /* no-op */
+    }
```
객체를 Json으로 직렬화하는 과정에서 빈 생성자가 필요할까요?

ProtoSeo replied:
직렬화 과정에서는 빈 생성자가 필요하지 않네요! 코드 수정했습니다!

sihyung92 commented on src/main/java/subway/dto/LineRequest.java:
```diff
@@ -2,22 +2,39 @@
+    private Long distance;
     private String color;
 
     public LineRequest() {
     }
```
기존에 있는 코드지만 jackson을 2.13버전 이후로 사용하는 코드의 경우 빈 생성자 없이도 객체를 생성할 수 있습니다.
https://github.com/FasterXML/jackson/wiki/Jackson-Release-2.13#no-constructor-deserializer-module

현재 버전의 스프링 부트는 2.13.5를 의존하고 있으므로 빈 생성자를 지우고, final을 이용해 사이드이펙트를 줄여볼 수 있겠습니다. 

<img width="627" alt="image" src="https://github.com/next-step/spring-subway/assets/51393021/c73d3cd7-3368-4da7-a724-059bca119b10">


\* 추신 : StationRequest 처럼 필드가 하나 있는 경우는 빈 생성자가 없으면 생성에 실패하는 버그가 있군요 -.-; 깃헙 등에서 리포트 된 이슈는 못 찾았는데 이 경우 때문에 StationRequest만 예외로 두기가 살짝 불편하긴 하네요. 저라면 StationRequest 빈 생성자는 내버려두고 주석으로 빈 생성자를 남긴 이유를 추가할거 같습니다.

ProtoSeo replied:
`StationRequest`를 제외한 기존의 Request 코드도 빈 생성자를 제거하고 final 키워드를 추가했습니다!

sihyung92 commented on src/main/java/subway/exception/SubwayException.java:
```diff
@@ -0,0 +1,8 @@
+package subway.exception;
+
+public class SubwayException extends RuntimeException {
```
비즈니스 커스텀 예외 👍

sihyung92 commented on src/main/java/subway/dto/LineWithStations.java:
```diff
@@ -0,0 +1,33 @@
+import subway.domain.Line;
+import subway.domain.Station;
+
+public class LineWithStations {
```
도메인 객체들을 다루고 있는 클래스인데요, dto 패키지에서 다루는 이유가 있나요?
비즈니스 로직이 없는건 Line도 마찬가지인데 왜 Line은 domain에 있고 이 객체는 dto 패키지에 있나요?

ProtoSeo replied:
처음 생각은 쿼리를 통해 `LineWithStations` 객체로 엔티티의 받아서, 이를 통해서 도메인의 객체들을 생성하기 때문에 엔티티와 도메인 계층 사이의 dto라고 생각했던 것 같은데 잘못 생각한 것 같아요.

sihyung92 commented on src/main/java/subway/domain/Section.java:
```diff
@@ -0,0 +1,122 @@
+
+public class Section {
+
+    private Long id;
+    private Long lineId;
+    private Long upStationId;
+    private Long downStationId;
+    private Long distance;
+
+    private Section() {
+        /* no-op */
+    }
```
불변을 포기하고 빈 생성자를 취했을 때 장점은 어떤게 있나요?

ProtoSeo replied:
빈 생성자를 가져가도 장점이 없을 것 같아요! 빈 생성자를 제거하고 final을 붙여서 불변으로 바꾸는 것이 더 좋을 것 같아요!

sihyung92 commented on src/main/java/subway/domain/Sections.java:
```diff
@@ -0,0 +1,121 @@
+                .collect(Collectors.toSet());
+    }
+
+    private int countByStationIds(final Long... targetIds) {
```
여러 ID가 많은 상황이니 targetStationIds라는 파라미터 명을 쓰면 더 가독성이 좋을 것 같습니다.

ProtoSeo replied:
아하! 메서드 명과 파라미터 명을 이어서 생각하면 충분할 거라고 생각했었어요! `targetStationsIds`처럼 더 명확하게 해주는 것도 좋을 것 같아요~

sihyung92 commented on src/main/java/subway/domain/Section.java:
```diff
@@ -0,0 +1,122 @@
+    }
+
+    private void validate(final Long upStationId, final Long downStationId, final Long distance) {
+        validateContainsUpStationAndDownStation(upStationId, downStationId);
+        validateDistanceNotNull(distance);
+        validateDistanceLessThanZero(distance);
```
Section 테스트를 보다 동일한 upStationId, downStationId를 가지는 섹션이 있어 확인해보니 동일한 상행 / 하행역을 가진 구간에 대한 validation이 없네요.

ProtoSeo replied:
아하! 이런 경우를 놓쳤네요! 테스트 및 Section에 validate 메서드 추가했습니다!

sihyung92 commented on src/test/java/subway/domain/SectionsTest.java:
```diff
@@ -0,0 +1,105 @@
+        final Optional<Section> lastSection = sections.findLastSection();
+
+        /* then */
+        assertThat(lastSection).isPresent();
```
isPresent가 아니라 new Section(3L, 3L, 4L, 3L) 과 동일성 비교를 하면 더 엄밀한 테스트가 될거 같아요. 
내부 로직이 알고리즘 같은 특성이 있는데, 이런 경우 잘 구성되어 있는 테스트가 코드 설명서 역할을 해줍니다. 그런데 코드 설명서가 엄밀하게 작성되어 있지 않다면  다른 개발자로 하여금 코드 기능에 대한 의심을 품게 되고, 이를 확인하기 위한 절차로 인해 생산성 저하로 이어집니다.

ProtoSeo replied:
맞네요! 명확하게 테스트를 하도록 수정했습니다!

sihyung92 commented on src/test/java/subway/domain/SectionsTest.java:
```diff
@@ -0,0 +1,105 @@
+        assertThat(sections.isEndStation(3L, 4L)).isFalse();
+        assertThat(sections.isEndStation(7L, 8L)).isFalse();
+    }
+}
```
EOL 이 있네요 ㅎ 사소하지만 - 가 거슬리므로 제거하는것도 좋습니다

ProtoSeo replied:
`EOL이 있다`라는 것이 Class 작성 완료 후 `}` 이후 개행이 없다는 의미일까요? 제가 이해한 의미와 다르다면 알려주시면 감사하겠습니다..!

sihyung92 replied:
아하 죄송합니다 제가 설명이 부족했습니다! EOL은 endOfLine의 준말로, 파일의 마지막 줄을 개행으로 끝내는 규칙을 말합니다(POSIX 계열의 운영체제 규칙)

잘 정리한 [블로그 링크](https://velog.io/@jangws/EOL%EC%9D%84-%EB%84%A3%EC%96%B4%EC%95%BC-%ED%95%98%EB%8A%94-%EC%9D%B4%EC%9C%A0%EC%99%80-git%EC%97%90%EC%84%9C-CRLF-EOL-%EC%B0%A8%EC%9D%B4%EB%A1%9C-%EC%9D%B8%ED%95%9C-%EB%AC%B8%EC%A0%9C-%ED%95%B4%EA%B2%B0-%EB%B0%A9%EB%B2%95)도 첨부합니다.

문제를 일으키는 경우가 많은 규칙은 아니지만, 깃헙 등에서 표시해주는게 보기싫으므로.. ㅎㅎ 그리고 긁어부스럼 일으킬 필요가 없으므로 지키는 것도 좋습니다.

sihyung92 commented on src/main/java/subway/domain/Sections.java:
```diff
@@ -0,0 +1,121 @@
+        final Map<Long, Section> upStationKeyMap = values.stream()
+                .collect(Collectors.toMap(Section::getUpStationId, section -> section));
+
+        final Section firstSection = findFirstSection().orElseThrow(IllegalArgumentException::new);
```
SubwayException을 사용하는 게 의도한 에러메시지를 내보내는 방법인거 같아요 ㅎ

ProtoSeo replied:
맞네요! 수정했습니다!

sihyung92 commented on src/main/java/subway/domain/Sections.java:
```diff
@@ -0,0 +1,121 @@
+package subway.domain;
+
+import java.util.*;
```
import ALL (asterisk import)는 보통 팀 컨벤션으로 금지하는 편입니다.
https://dev.to/kingori/import-5531

ProtoSeo replied:
아하! 수정했습니다. 최대한 구글 컨벤션을 따르려고 노력하고 있었는데, 구글 컨벤션에서도 금지하고 있었네요. 수정했습니다!

[No wildcard imports](https://google.github.io/styleguide/javaguide.html#s3.3.1-wildcard-imports)

sihyung92 commented on src/main/java/subway/domain/Sections.java:
```diff
@@ -0,0 +1,121 @@
+    }
+
+    public List<Long> getSortedStationIds() {
+        final Map<Long, Section> upStationKeyMap = values.stream()
+                .collect(Collectors.toMap(Section::getUpStationId, section -> section));
+
+        final Section firstSection = findFirstSection().orElseThrow(IllegalArgumentException::new);
+
+        final List<Long> result =
+                new ArrayList<>(List.of(firstSection.getUpStationId(), firstSection.getDownStationId()));
+        Section next = upStationKeyMap.get(firstSection.getDownStationId());
+
+        while (next != null) {
+            result.add(next.getDownStationId());
+            next = upStationKeyMap.get(next.getDownStationId());
+        }
+
+        return result;
```
생성시에 미리 정렬해놓고 upStationsIds + last downStationId를 보내는 것도 방법일거 같아요 ㅎ
장점 : values는 List타입이므로 정렬되어있을거라는 타 개발자의 기대감을 만족할 수 있음, 해당 메서드 같은 작업을 할 때 코드가 간편해짐
단점 : 정렬 비용이 기능 요구하는 런타임이 아니라 생성시점에 발생함 

그런데 객체 생성주기와 유즈케이스를 생각해보면 정렬 비용이 발생하는 시점이 이슈가 될거같진 않아서요, (동기로직을 가진 동일 메서드 내에서 함께 실행될 가능성이 높음) 저라면 생성시 정렬을 해놓을거 같긴 합니다.
(반영 하라는 내용은 아니니 넘기셔도 됩니다)

ProtoSeo replied:
좋은 아이디어라고 생각되어서 바로 반영했습니다!!

좋은 아이디어라고 생각한 이유는 지금 상행 종점역과 하행 종점역 뒤에 구간이 추가되는 경우 `O(n)`의 복잡도를 가지고 상행 종점역과 하행 종점역을 찾는데, 정렬이 되어 있다면, 상행 종점역과 하행 종점역을 찾을 때 `O(1)`의 시간복잡도로 찾을 수 있어서 장점이 더 크다고 생각했어요.

> 물론 Sections를 생성할 때 Map으로 정렬하므로 `O(n)`의 시간복잡도를 갖긴 하지만, 어차피 현재 `SectionService`에서 `createSection()` 메서드에서 매번 하행 종점역, 상행 종점역에 대한 검사를 하므로, 이 비용을 줄이는게 더 좋지 않을까? 라고 생각이 들었어요!

- 그리고 이 아이디어를 적용하면서, 구간 사이에 구간을 추가하는 경우, 잘못 추가되는 경우도 발견해서 수정했습니다.

sihyung92 replied:
넵 정말 최적화를 잘 하고 싶다면 정렬이 필요한 로직이 호출될 때 1회 수행하고, 그 결과값을 필드로 캐싱해놓는 방법도 있습니다 ㅎ
하지만 성능은 문제가 될때만 문제라서요, 이런 식의 처리가 괜히 가독성만 낮추는 결과로 이어질 때도 있습니다. 10만개 이상의 콜렉션이 예상되거나 할때는 문제가 되겠지만, 역이라는게 1000개가 넘기 힘드므로 성능보다 "읽기 좋은 코드"를 만드는게 더 나을지도요.

sihyung92 commented on src/main/java/subway/application/SectionService.java:
```diff
@@ -0,0 +1,82 @@
+
+        final Section lastSection = getLastSection(sections);
+        if (!lastSection.isSameDownStationId(downStationId)) {
+            throw new SubwayException("해당 노선에 일치하는 하행 종점역이 존재하지 않습니다.");
```
오류 상황에 대한 메시지를 좀더 풍부하게 만들면 미래에 자신에게 감사하게 됩니다.
```suggestion
            throw new SubwayException("해당 노선에 일치하는 하행 종점역이 존재하지 않습니다. 노선 ID : " + lineId + ", 하행 좀점역 ID : " + downStationId);
```

ProtoSeo replied:
좋은 것 같아요! 이렇게 하면 나중에 오류가 발생했을 때 좀 더 디테일하게 알 수 있을 것 같아요!

sihyung92 commented on src/main/java/subway/application/LineService.java:
```diff
@@ -42,6 +55,32 @@ public Line findLineById(Long id) {
     }
 
+    public LineResponse findLineAndStationsById(Long id) {
+        final List<LineWithStations> lineWithStations = findAllById(id);
```
애초에 LineWithSections로 조회해왔다면 아래 로직들이 필요없지 않을까? 생각했어요. lineDao.findAllById는 이미 Section에 join을 거는 쿼리인데 Section을 활용하고 있지 않아서요, 아래 로직들은 DB join을 손수 구현한듯한 느낌이었습니다

ProtoSeo replied:
그쵸,, 사실 이 코드 떄문에 제출하기 전에 도메인과 엔티티를 분리해야 하나? 하는 고민을 했었어요. 

여기서 `Section`으로 받지 않고, `Station`으로 받았던 이유도 같이 설명드리면, 현재 구현에서 `Section`에서 상행역, 하행역에 대한 ID값만 가지고, name에 대해서는 받지 않아요. 그렇기 때문에 `Section`으로 받으면 `Station`의 name을 어떻게 처리해야 하지? 라는 생각이 들었어요. 

그래서 `SectionEntity`를 만들어서 DB 테이블의 형태를 하고 `Section`도메인 그 안에는 `Stations` 객체를 가지고 있게 해야하나? 라는 생각을 많이 했던 것 같아요. (그리고서 평소에는 그냥 `Section` 도메인에서는 `Station`의 ID만 가지는 친구들만 들고 다니다가, `Station`의 name이 필요한 경우 `Station`의 ID와 name을 같이 들고 다니도록 하려고 생각했어요.)
- 근데 이렇게 하면 직접 구현이랑 차이가 거의 없어지는건 아닐까?? 하는 생각도 드네요..

그리고 Section과 join을 했던 이유는, `Line`과 `Station` 테이블 사이에서 `Line`에 대한 정보를 알기 위한 방법은 `Section`을 join하는 방법 뿐이어서 `Section`의 정보를 사용하지 않지만 `join`을 하게 되었던 것 같아요. (물론 `upStationId`, `downStationId` 는 사용해요.)

일단 이에 대한 고민은 차치하고 나서라도, 위 코드는 살짝 이상한 것 같기는 해서, `LineWithSection`으로 받아서, 라인 정보와 구간의 정보를 받고 `LineWIthSections`라는 객체를 생성해서, `List<LineWithSection>` 에서 `Sections`와 `Line`을 추출하도록 했어요. 그리고 `Station`의 name에 대한 문제는 `stationDao`에서 `where in` 절을 활용해서 `Sections`에서 가지고 있는 Station들을 불러오게 하고, 이 정보들을 `LineResponse`에서 합쳐서 반환하도록 구현 하도록 수정했습니다.

> 뭔가 지금 드는 생각으로는 JDBC를 좀 더 잘 썼으면 어땠을까? 아니면 쿼리를 좀 더 잘 짤 수 있었을까? 하는 생각도 같이 들긴 해요.

sihyung92 commented on src/main/java/subway/dao/LineDao.java:
```diff
@@ -5,6 +5,8 @@
 import org.springframework.jdbc.core.simple.SimpleJdbcInsert;
```
기존 Dao에 rowMapper에 final이 빠져있는데 추가해주셔용 ㅎ ㅎ

ProtoSeo replied:
저희가 기존에 작성된 코드는 최대한 안 건드리고 하다 보니 final 키워드가 빠져 있었네요! StationDao, LineDao에 빠진 final 키워드를 추가해주었습니다!

sihyung92 commented on src/main/java/subway/ui/SectionController.java:
```diff
@@ -0,0 +1,39 @@
+package subway.ui;
```
StationContoller 에 아래와 같은 처리가 있는데요, 기존코드지만 저는 매우 부적절한 처리같아요~~! 이유에 대해 생각해보시고 적절하게 수정해볼까요?

```java
    @ExceptionHandler(SQLException.class)
    public ResponseEntity<Void> handleSQLException() {
        return ResponseEntity.badRequest().build();
    }
```

ProtoSeo replied:
예외를 처리하는 일은 StationController 에서 할 일은 아닌 것 같아요! 그리고 SQLException은 서버의 오류로 발생해야 하는데, Bad Request 를 반환하는 것도 이상하네요!

RuntimeException을 받아서 처리하던 `RuntimeExceptionHandler`를 `CommonExceptionHandler`로 수정하고, 그 안에서 `SQLException`을 잡아서 처리할 수 있도록 수정했습니다!

sihyung92 reviewed (changes requested):
안녕하세요 승훈님? 리뷰어 성시형입니다.

전반적으로 기능을 잘 구현해주셨다는 생각이 들었습니다! 리뷰 한번 확인해주시면 감사하겠습니다.

### Q&A
- Q. repository 패턴을 접목했으면 어땠을까
- A. 레포지토리 패턴에서 추구하는 가치는 도메인레이어를 저장방식으로 부터 격리하는 게 장점입니다. (persistence layer가 DB든, 이벤트든, S3든 간에 상관없이 도메인을 재사용 가능) "Dao가 너무 많은 일을 하는게 걱정"인 상황에서 레포지토리 패턴을 적용한다고 이 문제가 해결될거 같지는 않습니다. 문제영역과 해결방안이 일치하지 않기 때문입니다. Dao가 많은 일을 한다고 여기신다면 어플리케이션 로직 쪽에서 더 많은 일을 수행하도록 리팩토링 해보시고, 이  방식에서 오히려 고통을 느끼신다면 "DAO가 많은 일을 하면 왜 안되는지"에 대해 고민하는게 우선이라는 생각이 듭니다.

---

- Q. 도메인과 엔티티
- A. 엔티티는 말하는 사람마다 멘탈모델이 다른 경향이 있습니다. 객체지향에서 얘기하는 식별자를 가진 객체를 엔티티라 부르는 사람이 있고, 테이블과 1대1관계를 맺는 테이블 모사 객체를 엔티티라고 부르는 사람도 있습니다. 적절히 소통해가면서 같은 개념을 이야기 하는게 맞는지 맞춰보면 되는거 같아요. 제가 생각하는 도메인 영역은 CPU랑 메모리를 활용해 해결해야 하는 영역을 의미합니다. 들어온 데이터랑 저장해놓은 데이터랑 나가는 데이터랑 잘 합쳐서 더하기를 하든 빼기를 하든 AI를 돌리든.. 우리가 원하는 액션을 만들어내기위한 "코드의 순서"를 도메인이라고 생각하긴 하는데요. 굳이 도메인 영역이라는 걸 만든 이유는 DB 기술이나 UI 기술은 맨날 바뀌니까 "데이터 처리해야하는 순서"는 오래오래 써먹고 싶어서 따로 분리했다고 생각합니다. 도움이 되는 답변이었는지는 모르겠네요 ㅎㅎ;

---

- Q. 도메인과 엔티티가 분리되지 않은 것 같아요
- A. 오 어쩌다보니 첫 답변에서 대답이 된거 같습니다. 2번 답변과 결합해보면 엔티티(저장되어야하는 데이터 형태 == 테이블 형태 객체를 말씀해주시는 것 같네요)가 따로 있었으면 문제(데이터베이스 의존 / 레이어 침범)가 없었을까? 일텐데요. 위에서 이야기한것처럼 효과가 클지는 의문이긴 하지만, 현재 어플리케이션에서 무언가 많이 하도록 노력하신거 같아서 분리해보는 것도 가능하겠다는 생각이 듭니다.

---

- Q. DAO와 repository의 추상화단계
- A. 저도 말씀해주신거랑 똑같이 생각합니다. 만약에 분리하신다면 repository 인터페이스는 도메인 영역에 두고 (도메인들이 높은 추상화단계를 의존하도록), 엔티티와 repository 구현체(JDBC 활용)는 persistence 영역으로 보내서 처리하는 방식이 있겠습니다.

---

- Q. 통합테스트를 TDD로 하기 어려워요
- A. 넵 저도 통합테스트를 TDD로 하는건 매우 힘들어보이는데요, 성공하게 만들려면 사실상 모든 코드를 다 만들어야 하니까요.. ㅎㅎ 테스트가 용이한 방향으로 코딩하다보면 좋은 코드가 만들어지는 경향이 있으니 TDD 하기 좋은 방식으로 쌓아보시는 것도 추천합니다. 즉 작은 도메인 코드부터 한 단계씩 키워나가는 방식으로요. 저는 만약 내가 하는 방식이 너무 고통스러우면 무언가 문제가 있다는 의미로 받아들입니다. 왜 힘들지? 의심을 해보는 것도 좋을것 같아요. 만약 현실적으로 접근한다고 한다면 API 테스트를 만들고 그걸 성공시키기 위한 mock controller를 만들어 일단 API 요청과 응답을 정해놓고, 그다음엔 도메인부터 쌓아나가 보는걸 추천드려보겠습니다. (첨부터 도메인부터 쌓으면 요구사항으로 받은 요청/응답이 안 맞는 경우도 생기더라구요)

ProtoSeo committed:
refactor: final 키워드 추가

ProtoSeo committed:
refactor: wildcard import 문 제거

ProtoSeo committed:
refactor: 빈 생성자 제거 및 final 키워드 추가

ProtoSeo committed:
refactor: Class 끝에 개행 추가

ProtoSeo committed:
refactor: SQLException Exception Handler에 추가

ProtoSeo committed:
test: ObjectMapper를 이용한 직렬화 역직렬화 테스트 추가

ProtoSeo committed:
refactor: Sections 리팩터링

- 테스트에만 사용하는 메서드 제거
- countByStationIds 매개변수 이름 변경
- 커스텀 예외로 수정

ProtoSeo committed:
refactor: 사용하지 않는 import문 제거

ProtoSeo committed:
feat: 상행역과 하행역이 같은 구간 생성 시 검증

ProtoSeo committed:
test: 상행역과 하행역이 같은 구간 생성 시 검증 추가로 인한 테스트 수정

ProtoSeo committed:
refactor: 빈 생성자 제거 및 final 키워드 추가

ProtoSeo committed:
refactor: 빈 생성자 접근 제어자 수정 및 설명 주석 추가

ProtoSeo committed:
refactor: 예외메시지 구체적으로 수정

ProtoSeo committed:
refactor: wildcard import문 제거

ProtoSeo committed:
refactor: 테스트 데이터 수정 및 테스트 수정

- 테스트 데이터, test 디렉터리로 이동
- 테스트 매직넘버 제거
- 공통적으로 사용하는 Section 추출

ProtoSeo committed:
refactor: import문 정렬 수정

ProtoSeo committed:
refactor: import문 정렬 수정 및 줄 바꿈

ProtoSeo committed:
refactor: 이름 변경 및 추가 리팩토링

- 빈 생성자 제거
- dto 패키지에서 domain 패키지로 이동
- Station 모음이 아니라, Section을 포함하도록 수정

ProtoSeo committed:
refactor: import문 정렬 변경

ProtoSeo committed:
feat: LineWithSections 를 활용하도록 수정

- LineWithStation을 LineWithSection으로 변경
- StationDao 에서 station id를 in절을 통해 station 목록을 받는 기능 추가
- LineResponse 에서 StationResponse 목록을 생성하도록 수정

ProtoSeo committed:
feat: Sections 생성 시 정렬하는 기능 추가

- Sections 생성 시 정렬하도록 기능 추가
- 사용하지 않는 메서드 제거
- Section 테스트 관련 수정

ProtoSeo committed:
fix: 구간 중간에 추가 시 남은 구간에 대한 잘못된 아이디 설정 관련 수정

sihyung92 commented on src/main/java/subway/application/SectionService.java:
```diff
@@ -63,20 +67,18 @@ private void validateDistance(final Long distance) {
     private void validateNotContainsBothStation(final Sections sections, final SectionRequest sectionRequest) {
-        if (sections.containsBoth(sectionRequest.getUpStationId(), sectionRequest.getDownStationId())) {
-            throw new SubwayException("상행 역과 하행 역이 이미 노선에 모두 등록되어 있습니다.");
+        final Long upStationId = sectionRequest.getUpStationId();
+        final Long downStationId = sectionRequest.getDownStationId();
+        if (sections.containsBoth(upStationId, downStationId)) {
```
한결 보기 좋네요 ㅎㅎ

sihyung92 commented on src/main/java/subway/dao/StationDao.java:
```diff
@@ -13,46 +14,58 @@
+            return Collections.emptyList();
+        }
+
+        String in = String.join(", ", Collections.nCopies(stationIds.size(), "?"));
+        final String sql = String.format("select id, name from STATION where id in (%s)", in);
```
비교적 가독성 있는 방법으로 기존 구현체인 NamedParameterJdbcTemplate를 활용한 방법도 있네요.  [링크](https://www.baeldung.com/spring-jdbctemplate-in-list#NamedParameterJdbcTemplate)

sihyung92 commented on src/main/java/subway/domain/LineWithSections.java:
```diff
@@ -0,0 +1,24 @@
+    }
+
+    public Line getLine() {
+        return values.get(0).getLine();
```
객체 생성시 validation을 해주지 않으면 해당 규칙을 믿고 쓸수 없지 않을까요? (LineWithSection은 모두 동일한 라인이다)

sihyung92 commented on src/main/java/subway/domain/Sections.java:
```diff
@@ -1,120 +1,123 @@
+    }
+
+    public boolean isFirstStation(final Long downStationId) {
+        final Long firstUpStationId = this.values.get(0).getUpStationId();
```
firstStation와 lastStation을 객체 생성시 필드로 미리 캐싱해두면

```java
this.values.get(0).getUpStationId();
```

이런 코드를 많이 줄일 수 있을 것 같아요.

sihyung92 commented on src/main/java/subway/exception/handler/CommonExceptionHandler.java:
```diff
@@ -8,11 +9,19 @@
+    @ExceptionHandler(SQLException.class)
+    @ResponseStatus(HttpStatus.INTERNAL_SERVER_ERROR)
+    public ExceptionResponse handleSQLException() {
+        return new ExceptionResponse(SQL_EXCEPTION_MESSAGE);
```
SQL 예외에 대해선 정보를 내려주지 않는 조치가 아주 좋네요 👍

sihyung92 commented on src/main/java/subway/domain/LineWithSections.java:
```diff
@@ -0,0 +1,24 @@
+import java.util.List;
+import java.util.stream.Collectors;
+
+public class LineWithSections {
```
객체 관계로 보자면 Line 안에 필드로 Sections가 들어가는 형태가 자연스럽지 않나요? 해당 객체는 객체지향 관점에선 다소 부자연스러워 보이네요 ㅎ

어려운 화두여서 미리 힌트를 드리자면, 이 부분이 객체와 DB테이블 간의 불일치가 일어나는 지점입니다. DB에서 쿼리를 쉽게 하려면 LineWithSections 같은 형태가 적절하고, 반대로 어플리케이션 로직을 쉽게 짜려면 Line이 Sections를 필드로 포함한 형태가 적절합니다. 

이런 불일치를 해소하기 위해 ORM (객체과 관계형 테이블간의 맵핑) 라이브러리들이 존재하는 것인데.. 현재 단계에서 어떤 식으로 미션을 구현해나가실지는 선택입니다 ㅎㅎ

sihyung92 commented on src/test/java/study/SerializeTest.java:
```diff
@@ -0,0 +1,45 @@
+
+    @Test
+    void serializeHasNotDefaultConstructor() {
+        final String message = "{\"value\":1}";
+
+        // 필드가 한 개인 경우, MismatchedInputException 발생
+        assertThatThrownBy(() -> om.readValue(message, OneRequest.class))
+                .isInstanceOf(MismatchedInputException.class);
```
변수명을 onlyOneFieldMessage (막네이밍 죄송..)등으로 했으면 필요없는 주석 아니었을까요?

sihyung92 commented on src/test/java/study/SerializeTest.java:
```diff
@@ -0,0 +1,45 @@
+import com.fasterxml.jackson.databind.exc.MismatchedInputException;
+import org.junit.jupiter.api.Test;
+
+public class SerializeTest {
```
다양한 학습 테스트를 구현해주셨네요 👍 대단하십니다

sihyung92 reviewed (approved):
안녕하세요 승훈님! 리뷰어 성시형입니다. 
구현 잘 해주셨네요 ㅎ 학습테스트가 인상적이었습니다.
몇가지 코멘트와 리뷰 남겼는데, 확인해주셔요!
코드는 머지하고 다음 단계 때 뵙겠습니다
